### ISSUE TITLE
Performance and memory consumption problems with XMLProvider

### ISSUE THREAD

**Issue by Thorium** [fsharp/FSharp.Data#975]
I'm using XMLProvider and the files are local.
- After developing for 10 mins, my Visual Studio consumes over 2GB of memory.
- If I change the XML file content, the Visual Studio freezes totally.

**Comment by giacomociti** [comment]
I'm facing the same problem tweaking the XML provider to support xsd.
Schemas (instead of xml samples) get parsed again and again on every keystroke, making Visual Studio unusable when dealing with multiple big schemas.
Things get a lot better caching instances of `ProvidedTypeDefinition`.
I'm not completely happy with my [current solution](https://github.com/giacomociti/FSharp.Data.Xsd/commit/86190b95c810a39b1bb0ed9ecaf0cb5f06df6c81), I'm afraid I'm leaking memory and also that changes to files (albeit unlikely to happen) are ignored. But I hope it's just a matter of implementing caching right for the `buildTypes` function.

**Comment by Thorium** [comment]
Any news on this one?

**Comment by giacomociti** [comment]
I haven't looked into this again, but surely the memory leak is due to the cache not purging expired items (your PR should fix this).

**Comment by Thorium** [comment]
I was thinking if @ovatsus would have checked the parsing-again-and-again problem... :-)

**Comment by giacomociti** [comment]
lurking around I noticed an old warning of @dsyme about caching in aed22f01144f16a5b689c0408e8af8dbd4bd2163
Out of curiosity, I just commented out writing to `providedTypesCache `at line 305 of [Helpers.fs](FSharp.Data/src/CommonProviderImplementation/Helpers.fs).
Playing a bit in visual studio with a few big files, it seems that memory is released far more often. 

Mine is not a decent analysis (just looked at windows activity monitor) and I'm still unaware of how type providers work and the related object's lifecycle. But at least this may be a hint useful to someone able to make such an analysis
